### ISSUE TITLE
Empty Variables 

### ISSUE THREAD

**Issue by c0cky** [felixfbecker/vscode-php-debug#43]
PHP version:  5.4.24
XDebug version:  2.2.3
Adapter version:  

<img width="398" alt="screen shot 2016-06-15 at 3 26 18 pm" src="https://cloud.githubusercontent.com/assets/3203697/16094326/ba72602c-330d-11e6-8ded-c176f2c71789.png">

None of these variables should be none

**Comment by felixfbecker** [comment]
Can you please post your log files from setting `"log": true` and setting `xdebug.remote_log` in php.ini?

**Comment by c0cky** [comment]
php.ini on VM:
```
[xdebug]
xdebug.remote_enable=1
xdebug.remote_handler=dbgp
xdebug.remote_host=10.0.2.2
xdebug.remote_port=9000
xdebug.remote_autostart=1
xdebug.remote_log=/var/log/xdebug.log
```

log files from log: true

```
<- launchResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 2,
  command: 'launch',
  success: true }
new connection 1
<- threadEvent
ThreadEvent {
  seq: 0,
  type: 'event',
  event: 'thread',
  body: { reason: 'started', threadId: 1 } }
-> threadsRequest
{ type: 'request', seq: 3, command: 'threads' }
<- threadsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 3,
  command: 'threads',
  success: true,
  body: { threads: [ Thread { id: 1, name: 'Request 1 (4:51:33 PM)' } ] } }
<- initializedEvent
InitializedEvent { seq: 0, type: 'event', event: 'initialized' }
-> setBreakpointsRequest
{ type: 'request',
  seq: 4,
  command: 'setBreakpoints',
  arguments: 
   { source: { path: '/usr/share/www/src/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php' },
     lines: [ 34 ],
     breakpoints: [ { line: 34 } ] } }
<- setBreakpointsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 4,
```

its also getting hung up in the browser and the page loads infinitely.

**Comment by felixfbecker** [comment]
@c0cky I need the content of `/var/log/xdebug.log`

**Comment by Dreykodoon** [comment]
I am having the same problem as @c0cky.

PHP version: 5.5.9-1ubuntu4.17
XDebug version: 2.2.3

The output of the vscode debug console:
[vscode_debug.txt](https://github.com/felixfbecker/vscode-php-debug/files/317682/vscode_debug.txt)

The output of the xdebug remote log file:
[xdebug_remote.txt](https://github.com/felixfbecker/vscode-php-debug/files/317683/xdebug_remote.txt)

Hope this helps.

**Comment by felixfbecker** [comment]
@Dreykodoon thanks for the log file. What is the name of the variable that should not be empty?
@c0cky does this happen only with properties or also with variables?

**Comment by felixfbecker** [comment]
@c0cky I found it in your logs, the variable in question in `$txt`, right? The thing is, it comes back from XDebug empty:
```xml
<property name="$txt" fullname="$txt" address="140463401248360" type="string" size="13" encoding="base64"><![CDATA[]]></property>
```
Maybe this is a bug in XDebug.

**Comment by felixfbecker** [comment]
@c0cky Posting your screenshot from Gitter here aswell:
![image](https://cloud.githubusercontent.com/assets/10532611/16107785/e691170c-339e-11e6-9d49-8bcd339fd81e.png)

**Comment by felixfbecker** [comment]
I cannot reproduce this on my machine. There are tests for the variables, [this file](https://github.com/felixfbecker/vscode-php-debug/blob/master/testproject/variables.php) is used to test variables output [here](https://github.com/felixfbecker/vscode-php-debug/blob/master/src/test/adapter.ts#L374), and the tests are run on PHP 5.4 in Travis [here](https://travis-ci.org/felixfbecker/vscode-php-debug/builds/137720388).

Could you please try out the variable.php file and see if you can reproduce the issue with that file?

**Comment by c0cky** [comment]
@felixfbecker yeah I will. Here is the log for when the debugger hangs up the browser

```
Log opened at 2016-06-15 20:51:37
I: Connecting to configured address/port: 10.0.2.2:9000.
I: Connected to client. :-)
-> <init xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" fileuri="file:///usr/share/www/base_includes/config/php/redventures_init.php" language="PHP" protocol_version="1.0" appid="10206"><engine version="2.2.3"><![CDATA[Xdebug]]></engine><author><![CDATA[Derick Rethans]]></author><url><![CDATA[http://xdebug.org]]></url><copyright><![CDATA[Copyright (c) 2002-2013 by Derick Rethans]]></copyright></init>

<- feature_set -i 1 -n max_depth -v 1
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="feature_set" transaction_id="1" feature="max_depth" success="1"></response>

<- feature_set -i 2 -n max_children -v 9999
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="feature_set" transaction_id="2" feature="max_children" success="1"></response>

<- feature_set -i 3 -n max_data -v 0
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="feature_set" transaction_id="3" feature="max_data" success="1"></response>

<- breakpoint_list -i 10
<- breakpoint_list -i 6
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="6"><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102080003"></breakpoint></response>

-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="10"><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102130001"></breakpoint><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102130002"></breakpoint><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102130003"></breakpoint></response>

<- breakpoint_list -i 8
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="8"><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102070005"></breakpoint><breakpoint type="line" filename="file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php" lineno="34" state="enabled" hit_count="0" hit_value="0" id="102070004"></breakpoint></response>

<- breakpoint_list -i 4
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="4"></response>

<- breakpoint_set -i 11 -t line -f file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php -n 34
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="11" id="102130004"></response>

<- breakpoint_set -i 7 -t line -f file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php -n 34
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="7" id="102080004"></response>

<- breakpoint_set -i 9 -t line -f file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php -n 34
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="9" id="102070006"></response>

<- breakpoint_set -i 5 -t line -f file:///usr/share/www/dob/libs/libraries/dob/orderprocess/action/dob_orderprocess_action_flightinfopost.php -n 34
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="5" id="102060005"></response>```

**Comment by G3z** [comment]
this does happen to me when debugging php in Vagrant but not with php in my local machine

here some info about the vagrant machine:
```
vagrant@vagrant-ubuntu-trusty-64:~$ php -v
PHP 5.5.9-1ubuntu4.16 (cli) (built: Apr 20 2016 14:31:27)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
    with Xdebug v2.2.3, Copyright (c) 2002-2013, by Derick Rethans
```
```
[xdebug]
zend_extension=xdebug.so
xdebug.remote_autostart=on
xdebug.remote_enable=on
xdebug.remote_enable=1
xdebug.remote_port=9001
xdebug.profiler_enable=0
xdebug.remote_connect_back=1
xdebug.profiler_enable_trigger=1
xdebug.profiler_output_dir=/tmp
xdebug.profiler_output_name = "cachegrind.out.%t-%s"
xdebug.idekey="xdebug"
xdebug.remote_handler=dbgp
xdebug.remote_host=192.168.33.1
xdebug.remote_mode=req
xdebug.max_nesting_level=256
```

**Comment by G3z** [comment]
updating xdebug to 2.4.0 fixed the issue for me

**Comment by c0cky** [comment]
@G3z how did you update xdebug on vagrant?

**Comment by G3z** [comment]
my machine is a `Ubuntu 14.04.4 LTS`
that said this is the command sequence (although not optimal because it brutally replaces old xdebug )
```bash
# get inside the machne
$ vagrant ssh
# get xdebug 2.4
$ wget -O xdebug-2.4.0.tgz http://xdebug.org/files/xdebug-2.4.0.tgz
# build it
$ tar -xvzf xdebug-2.4.0.tgz
$ cd xdebug-2.4.0
$ phpize
$ ./configure
$ make
# replace the old one
$ sudo cp modules/xdebug.so /usr/lib/php5/20121212/
# check 
$ php -v
# resart apache
$ sudo service apache2 restart
```

i hope it helps

**Comment by felixfbecker** [comment]
@3thanZ I'm sorry to hear that it isn't working, but really there is nothing I can do about it. See my [previous comment](https://github.com/felixfbecker/vscode-php-debug/issues/43#issuecomment-226401670), the error is not on my side, XDebug actually sends me an empty string for the variable. It clearly is an XDebug bug that is fixed in newer releases.

**Comment by 3thanZ** [comment]
@felixfbecker Thanks for the speedy response. I've attached the xdebug log when running Eclipse and another when running VS Code with your extension. Both when breakpoints are set on the same line.  They do look different but I have no idea how to make sense of it. 

Looking forward to your thoughts and have a good day ahead.

[xdebug-logs.zip](https://github.com/felixfbecker/vscode-php-debug/files/321813/xdebug-logs.zip)

**Comment by felixfbecker** [comment]
@3thanZ Can you tell me the name of the variable that is empty?

**Comment by 3thanZ** [comment]
@felixfbecker The variable name is $lang. I've attached the screenshots of both VSCode and Eclipse.

[screenshots.zip](https://github.com/felixfbecker/vscode-php-debug/files/321836/screenshots.zip)

**Comment by felixfbecker** [comment]
I think I know why it fails, I set `max_data` to `0` because zero means unlimited. Maybe in older versions it was treated as literally zero. I added XDebug 2.2 to the build matrix but there are a lot of test failures because it didn't report user defined constants for example.

**Comment by 3thanZ** [comment]
@felixfbecker Nice .... but could only understand the first 2 sentences ;) Possible to rephrase it for a noob like me? Hahaha. Thanks.

**Comment by felixfbecker** [comment]
@3thanZ See here: https://travis-ci.org/felixfbecker/vscode-php-debug
I added XDebug 2.3.3. The test for strings is indeed failing, see [here](https://travis-ci.org/felixfbecker/vscode-php-debug/jobs/138596354#L924). But unfortunately also other tests are failing that I have to fix.

**Comment by felixfbecker** [comment]
Here is the issue at XDebug btw: https://bugs.xdebug.org/bug_view_page.php?bug_id=00000979
It was fixed in 2.2.4

**Comment by 3thanZ** [comment]
@felixfbecker Let me know what I can do to help ;)

**Comment by felixfbecker** [comment]
@3thanZ This should fix your issue. I will do a release once I fixed all the other test failures.
[Uploading php-debug-1.9.1.vsix.zip…]()

**Comment by 3thanZ** [comment]
Thanks @felixfbecker .... Just out of curiosity, why not make `max_data` a setting that the user can configure?

**Comment by 3thanZ** [comment]
@felixfbecker .... Tested it and it works! Thanks a bunch for fix.

**Comment by felixfbecker** [comment]
@3thanZ You can set it in php.ini, but the less config required the better. I have added checks against the XDebug version.

**Comment by stefanorosanelli** [comment]
@felixfbecker I had the same problem, I can confirm that with your fix it works now
thanks

**Comment by felixfbecker** [comment]
Included in 1.9.2
